aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
   return at::any(self._values());
 }
 
+Tensor bmm_sparse_cpu(const SparseTensor& self, const Tensor& mat2) {
+  Tensor result = at::empty({}, mat2.options());
+  return bmm_out_sparse_cpu(result, self, mat2);
+}
+
+// Search a sorted strided array for the rightmost instance of a value.
+// Array must be sorted from lowest to highest
+template<typename scalar_t>
+scalar_t binary_search_strided_rightmost(scalar_t search_val, scalar_t* sorted_arr, int64_t sorted_arr_stride, int64_t length, bool* found) {
+  if (length == 0) {
+    *found = false;
+    return -1;
+  }
+
+  int64_t left_ind = 0;
+  int64_t right_ind = length - 1;
+  int64_t mid_ind;
+  bool done_searching = false;
+
+  while (!done_searching) {
+    mid_ind = (left_ind+right_ind) >> 1;
+    scalar_t mid_val = sorted_arr[mid_ind*sorted_arr_stride];
+
+    if (mid_val > search_val) {
+      right_ind = mid_ind-1;
+    } else if((mid_val == search_val) && (
+      (mid_ind == length - 1) || (sorted_arr[(mid_ind+1)*sorted_arr_stride] != search_val)
+    )) {
+      done_searching = true;
+      *found = true;
+    } else {
+      left_ind = mid_ind+1;
+    }
+
+    if (left_ind > right_ind) {
+      done_searching = true;
+      *found = false;
+      mid_ind = -1;
+    }
+  }
+
+  return mid_ind;
+}
+
+Tensor& bmm_out_sparse_cpu(Tensor& result, const SparseTensor& self, const Tensor& mat2) {
+  TORCH_CHECK(!mat2.is_sparse(), "bmm_sparse: Tensor 'mat2' must be dense");
+
+  TORCH_CHECK(self.dense_dim() == 0, "bmm_sparse: Tensor 'self' must have 0 dense dims, but has ", self.dense_dim());
+  TORCH_CHECK(self.sparse_dim() == 3, "bmm_sparse: Tensor 'self' must have 3 sparse dims, but has ", self.sparse_dim());
+  TORCH_CHECK(mat2.dim() == 3, "bmm_sparse: Tensor 'mat2' must have 3 dims, but has ", mat2.dim());
+
+  TORCH_CHECK(self.size(0) == mat2.size(0), "bmm_sparse: 'self.size(0)' and 'mat2.size(0)' must match");
+  TORCH_CHECK(self.size(2) == mat2.size(1), "bmm_sparse: 'self.size(2)' and 'mat2.size(1)' must match");
+
+  result.resize_({self.size(0), self.size(1), mat2.size(2)});
+
+  if (self._nnz() == 0) {
+    result.zero_();
+    return result;
+  }
+
+  // First need to coalesce to get all of the first dimension indices
+  // in order since we'll be sending each matrix into the MM operation
+  SparseTensor self_coalesced;
+
+  if (!self.is_coalesced()){
+    self_coalesced = self.coalesce();
+  } else {
+    self_coalesced = self;
+  }
+
+  int64_t nnz =        self_coalesced._nnz();
+  LongTensor indices = self_coalesced._indices();
+  Tensor values =      self_coalesced._values();
+
+  LongTensor indices_dim0 = indices[0];
+  int64_t indices_dim0_stride = indices_dim0.stride(0);
+  int64_t* indices_dim0_ptr = indices_dim0.data_ptr<int64_t>();
```

Comment:
As the dimensionality and data type are known statically, you can use `TensorAccessor` for a much nicer, operator[] based API here.

Reply:
Done